Lib/test/test_patma.py

```diff
@@ -3056,6 +3057,66 @@ class Keys:
         self.assertIs(z, None)
 
 
+class TestTracing(unittest.TestCase):
+
+    def _test_trace(self, func, expected_linenos, *f_args):
+        def trace(frame, event, arg):
+            if frame.f_code.co_name == func.__name__:
+                relative_lineo = frame.f_lineno - func.__code__.co_firstlineno
+                self.assertIn(relative_lineo, expected_linenos)
+            return trace
+
+        sys.settrace(trace)
+        func(*f_args)
+        sys.settrace(None)
```

Comment:
The way this test is currently designed, we could fail to catch the case where we expect to see a line, but don't. Maybe it can be rewritten like this to test that all the expected lines are covered (no more, no less)?

```suggestion
        actual_linenos = set()
        def trace(frame, event, arg):
            if frame.f_code.co_name == func.__name__:
                relative_lineno = frame.f_lineno - func.__code__.co_firstlineno
                actual_linenos.add(relative_lineno)
            return trace

        sys.settrace(trace)
        func(*f_args)
        sys.settrace(None)
        self.assertSetEqual(actual_linenos, expected_linenos)
```

Reply:
I like this alot better - thanks!